homeassistant/components/modbus/modbus.py

```diff
@@ -56,9 +74,14 @@ def modbus_setup(
         for component, conf_key in (
             (CONF_CLIMATE, CONF_CLIMATES),
             (CONF_COVER, CONF_COVERS),
+            (CONF_BINARY_SENSOR, CONF_BINARY_SENSORS),
+            (CONF_SENSOR, CONF_SENSORS),
+            (CONF_SENSOR, CONF_BIT_SENSORS),
```

Comment:
traditionally every platform have its own domain and own file, but that is not a strong demand it depends a lot more on the added complexity.

Reply:
Lmk your best judgment on that, I can split it, though the file size, seems, manageable.
Also, I assume I can't make (and it make no sense in this case) a new domain for, let's say, a bit_sensor.

Reply:
Oh, cool, for some reason I thought CONF_BINARY_SENSOR is a built-in constant.

Reply:
Hi @janiversen, maybe you can help me with the issue I can't find any documentation about:
I added the following into the `const.py`
```
# bit_switch.py
CONF_BIT_SWITCH = "bit_switch"
CONF_BIT_SWITCHES = "bit_switches"
```
then in a `modbus.py` I have: 
```
        # load platforms
        for component, conf_key in (
            ....
            (CONF_BIT_SWITCH, CONF_BIT_SWITCHES),
        ):
            if conf_key in conf_hub:
                load_platform(hass, component, DOMAIN, conf_hub, conf_hub)
```
Also, I have `bit_switch.py` which exposes `DOMAIN` variable eq to `modbus` and has `async_setup_platform` function in it. Pretty much the same set up as `switch.py` has, except file name is `bit_switch`
When I try to run a test: 
```
    BIT_SWITCH_DOMAIN = "bit_switch"
    switch_name = "modbus_test_bit_switch"
    state = await base_test(
        hass,
        {
            CONF_NAME: switch_name,
            CONF_INPUT_TYPE: CALL_TYPE_REGISTER_INPUT,
            CONF_ADDRESS: 1234,
            CONF_SLAVE: 1,
            CONF_COMMAND_BIT_NUMBER: 5,
        },
        switch_name,
        BIT_SWITCH_DOMAIN,
        CONF_BIT_SWITCHES,
        CONF_REGISTERS,
        regs,
        expected,
        method_discovery=True,
        scan_interval=5,
    )
    assert state == expected
```
I'm getting 
`ERROR    homeassistant.setup:setup.py:138 Setup failed for bit_switch: Integration not found.`

What I'm missing? I assume I have to register `bit_switch` somehow. Your help would be very appreciated.

Reply:
Extracting bit sensor/switch into separate files will require extracting the base switch/sensor to the separate file as well to avoid the import cycle. That way we will have a bunch of short files. If you prefer that way, LMK, I will extract `base_swith` `base_sensor` `register_sensor` `bit_sensor` `register_swith` `bit_switch` `coil_swith` and leave schemas and `async_setup_platform` in the `sensor` and `switch`

Reply:
No it will not !! see how we use base switch today. The idea is to keep functionality that belong together in the same file. Therefore sensor and base_sensor belong in one file, whereas specializations of sensor belongs in other files. Same goes for switch and base_switch. 

Cover and Climate are good examples of how to use a base class.

Do NOT touch the schemas in switch/sensor etc. those are the old configuration style and frozen. All new configuration changes goes into __init__.py

So all in all do not touch sensor.py or switch.py that should not be needed to implement bit_sensor.

The reasons for this grouping are
- coverage tell more precisely where tests are missing
- Any problem in bit_sensor that are not in other sensors are to be solved in that file.

Reply:
If you look from the HA side as opposed to the device side a bit_sensor is actually a binary_sensor and a bit_switch is just a switch. So the domains are BINARY_SENSOR and SWITCH.